self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoicesTestCase.java

```diff
@@ -134,33 +138,153 @@ public void paysInvoice(){
+            project.repoFullName(),
+            contributor.username(),
+            contributor.provider(),
+            BigDecimal.ONE,
+            Contract.Roles.DEV
+        );
+        storage.contracts().addContract(
+            project.repoFullName(),
+            contributor.username(),
+            contributor.provider(),
+            BigDecimal.ONE,
+            Contract.Roles.ARCH
+        );
+        final Contract.Id contractIdOne = new Contract.Id(
+            project.repoFullName(),
+            contributor.username(),
+            contributor.provider(),
+            Contract.Roles.DEV
+        );
+        final Contract.Id contractIdTwo = new Contract.Id(
+            project.repoFullName(),
+            contributor.username(),
+            contributor.provider(),
+            Contract.Roles.ARCH
+        );
+        storage.invoices().add(this.mockAssignedTask(
+            "123", storage, project, contributor, Contract.Roles.DEV),
+            Duration.ofHours(1));
+        storage.invoices().add(this.mockAssignedTask(
+            "124", storage, project, contributor, Contract.Roles.ARCH),
+            Duration.ofHours(2));
 
+        storage.invoices().ofContract(contractIdTwo).isPaid(1);
     }
 
+    /**
+     * Throws IllegalStateException if invoice tasks are not part of contract.
+     */
+    @Test(expected = IllegalStateException.class)
+    public void throwsIfInvoiceTasksNotPartOfContract(){
+        final Storage storage = new InMemory();
+        final ProjectManager projectManager = storage.projectManagers()
+            .pick(Provider.Names.GITHUB);
+        final Project project = storage
+            .projects()
+            .register(
+                this.mockRepo("john/test", Provider.Names.GITHUB),
+                projectManager
+            );
+        final Contributor contributor = storage
+            .contributors()
+            .register("mihai", Provider.Names.GITHUB);
+        //adding two contracts
+        storage.contracts().addContract(
+            project.repoFullName(),
+            contributor.username(),
+            contributor.provider(),
+            BigDecimal.ONE,
+            Contract.Roles.DEV
+        );
+        storage.contracts().addContract(
+            project.repoFullName(),
+            contributor.username(),
+            contributor.provider(),
+            BigDecimal.ONE,
+            Contract.Roles.ARCH
+        );
+        final Contract.Id contractIdTwo = new Contract.Id(
+            project.repoFullName(),
+            contributor.username(),
+            contributor.provider(),
+            Contract.Roles.ARCH
+        );
+        storage.invoices().add(this.mockAssignedTask(
+            "123", storage, project, contributor, Contract.Roles.DEV),
+            Duration.ofHours(1));
+        storage.invoices().add(this.mockAssignedTask(
+            "124", storage, project, contributor, Contract.Roles.ARCH),
+            Duration.ofHours(2));
```

Comment:
@criske Shouldn't this test fail here or when adding the first task above? Why does it fail only when trying to read the Task from teh contract (the last line  bellow)?

Reply:
Will not fail because we adding two tasks each linked with a different contract. (one contract has DEV role other ARCH, diff keys basically).

The last line will fail becase we're trying to get tasks of an invoice that is not part of the contract.
Invoice with id 1 is part of the contractOne not contractTwo.